### PR TITLE
Migrate unit tests to satysfi-test

### LINKED ISSUE #142: Try migrating unit tests to satysfi-test
# Background
At present, we have unit tests and regression tests to keep quality. We write unit tests by writing all test results to output and compare them with snapshots because we don't have unit testing frameworks for SATySFi.

I published the very first version of [satysfi-test](https://github.com/zeptometer/satysfi-test). I consider we can try it out, and replace all unit tests to ones with satysfi-test if it works well.

# Goal
We should have better unit tests using satysfi-test.

### PR DESCRIPTION
**Pull request by zeptometer**
This PR migrates regression tests from `generic.saty` to unit tests with [satysfi-test](https://github.com/zeptometer/satysfi-test). This closes #142.

# What this PR does
This PR introduces a new directory `/test` with test definitions. The top-level file `main.test.saty` loads all test sources. `run-test.sh` compiles the whole source in text-mode, and output `report.txt`, including test result. It runs in CI to ensure that satysfi-base works up to expectation, which you can find `.github/workflows/ci.yml`.

# Why to migrate to satysfi-test
The regression tests from `generic.saty` make snapshots from compiler output, which is not very readable and maintainable. Unit tests with satysfi-test make tests organized, clear and maintainable. It also allows test-first development: It will be beneficial for future development.

### CODE REVIEW

File: src/ord.satyg [hunk 1]
```diff
@@ -6,6 +6,29 @@
 
 type ordering = Lt | Gt | Eq
 
+module Ordering : sig
```
Comment (zeptometer) [inline src/ord.satyg:9]: Added eq for `ordering`. I'm a bit afraid that `Ord` and `Ordering` are confusing. Any suggestion?

File: src/tuple.satyg [hunk 2]
```diff
@@ -11,9 +12,18 @@ module Pair : sig
   val fst : ('a * 'b) -> 'a
   val snd : ('a * 'b) -> 'b
   val swap : ('a * 'b) -> ('b * 'a)
+
+  val eq : 'a Eq.t -> 'b Eq.t -> ('a * 'b) Eq.t
```
Comment (nyuichi) [inline src/tuple.satyg:16]: adding `implicit`s would be nice (for the documenting purpose)
```suggestion
  val eq : ('a Eq.t) implicit -> ('b Eq.t) implicit -> (('a * 'b) Eq.t) implicit
```

File: test/promise.test.satyg [hunk 3]
```diff
@@ -0,0 +1,20 @@
+@require: test/test
+@require: test/expect
+
+@import: ../src/base0
+@import: ../src/ref
+@import: ../src/promise
+
+let promise-test-cases = open Test in
+  describe `promise module` [
+    it `evaluate lazily` (fun () -> (
```
Comment (zeptometer) [inline test/promise.test.satyg:10]: 1. The original test uses `Debug.log` to check the order of side effects, but this is not feasible with the satisfy-test. Instead, this test case uses `ref` to make side effects.
2. The original test has additional cases that I did not find helpful. That's why they are omitted in this test case.

File: test/regex.test.satyg [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  describe `RegExp module` [
+    % TODO: Do we need this?
+    % let () = RegExp.test (RegExp.of-string `ab.*ef`) `abcdef` |> String.of-bool |> Debug.log in
+    % let () = exec (compile (seq (many1 char-a) (many1 char-b))) `aabbb` |> String.of-bool |> Debug.log in
+    % let () = exec re-example1 `aabbb` |> String.of-bool |> Debug.log in
+    it `test "(a+)(b+)"` (fun () -> (
+      test (seq (group (many1 char-a)) (group (many1 char-b))) `aabbb`
+      |> Expect.is-true
+    ));
+    % TODO GH-142: Is this right description?
```
Comment (zeptometer) [inline test/regex.test.satyg:21]: I try to make descriptions from test cases, but I need more confidence with some of them. Need help from @nyuichi

File: test/float.test.satyg [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        |> flleq [2.; 4.; 6.; neg 2.; neg 4.; neg 6.; 6.])
+      && (l |> List.map floor
+        |> flleq [2.; 3.; 5.; neg 3.; neg 4.; neg 6.; 6.])
+      && (l |> List.map ceil
+        |> flleq [3.; 4.; 6.; neg 2.; neg 3.; neg 5.; 6.])
+      && (l |> List.map truncate
+        |> flleq [2.; 3.; 5.; neg 2.; neg 3.; neg 5.; 6.])
+      |> Expect.is-true
+    ));
+    it `virtually test exp2i` (fun () -> (
```
Comment (zeptometer) [inline test/float.test.satyg:104]: What would be a good description? @nyuichi

File: test/fn.test.satyg [hunk 6]
```diff
@@ -0,0 +1,17 @@
+@require: test/test
+@require: test/expect
+
+@import: ../src/fn
+
+let fn-test-cases = open Test in
+  describe `Fn module` [
+    it `make recursive function with fix` (fun x -> (
```
Comment (zeptometer) [inline test/fn.test.satyg:8]: The original test case calls `Debug.log` for each recursion, but it is not feasible in satysfi-test. That's why I write this case in this way.

**Review by nyuichi — APPROVED** [review]

**Pull request merged**